### PR TITLE
Cache Rust dependencies

### LINKED ISSUE #18: Improve caching in the Docker build
The Docker image currently takes quite a long time to build (~20 minutes in GitHub Actions). A large part of that time is spent running `cargo`, fetching the crates.io index and downloading packages:

https://github.com/praetorian-inc/noseyparker/blob/9438ab22c3d3996861cde683f85a2a08bf89e484/Dockerfile#L29-L30

Is it possible and feasible to expose the contents of the GitHub Actions cache to `docker build` to speed it up?

### PR DESCRIPTION
**Pull request by grawin**
Fixes #18

Added stage to Dockerfile for fetching and building Rust dependencies based on project's Cargo files. Creates stub files to facilitate build then in following stage copies full project and updates file timestamps to cause a compile on the actual project. Once the build is cached subsequent code changes will only impact the project's build stage. It's a little clunky, but unfortunately Cargo currently does not have a "fetch / build dependencies only" command.

Tested on my fork with CI builds. Initial Docker CI build takes ~15 minutes then subsequent build with 1 line code change took ~2 minutes.

Misc. comment: Could consider using a distroless image such as gcr.io/distroless/cc-debian11 in place of debian:11-slim for a smaller final run-time image.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,38 +1,76 @@
...
+COPY ["Cargo.toml", "Cargo.lock",  "./"]
+
+# Make directory structure
+RUN mkdir -p  ./src/bin/noseyparker &&\
+    mkdir -p ./benches &&\
+    # Create stub files for build
+    touch ./benches/microbench.rs ./src/lib.rs &&\
+    echo "fn main() {}" > ./src/bin/noseyparker/main.rs &&\
+    # Run the build
+    cargo build --bin noseyparker --release
```
Comment (bradlarsen) [inline Dockerfile:53]: I _think_ I understand what's going on here: all this song and dance is to convince `cargo` to build just the dependencies of the project, essentially, which it doesn't support natively. If I understand that right, please add a 1-2 sentence comment noting what's going on, for future readers.

Is it actually necessary to create dummy `benches/microbench.rs` and `src/lib.rs`, or is it sufficient to only create the dummy `src/bin/noseyparker/main.rs`?

File: Dockerfile [hunk 2]
```diff
@@ -1,38 +1,76 @@
...
+
+WORKDIR "/noseyparker"
+
 COPY . .
-# build against from-source vectorscan
-ENV HYPERSCAN_ROOT "/vectorscan/build"
-# XXX it would be nice if this could store crates.io index and dependency builds in the Docker cache
+
+# Update file timestamps
+RUN touch ./benches/microbench.rs ./src/lib.rs ./src/bin/noseyparker/main.rs
```
Comment (bradlarsen) [inline Dockerfile:65]: Is it actually necessary to update the timestamps here?

**Review by bradlarsen — CHANGES_REQUESTED** [review]
Thanks @grawin! This is cool. I have a couple questions about what's going on here before I merge it though.

And yes, good suggestion about using a distroless image for the final image -- all it needs is glibc, libstdc++, maybe one or two other shared libs. It could be shrunk down.

File: Dockerfile (hunk 1 above)
Comment (grawin) [reply Dockerfile]: Added comments in an attempt to describe what is going on. Those three files are needed to do a build even if you specify just the bin target unfortunately.

File: Dockerfile (hunk 2 above)
Comment (grawin) [reply Dockerfile]: Yes, the Rust compiler must use timestamps as part of how it figured out what work it needs to do. The timestamps will always be older in the real code versus the temporary stub files so they need to be modified.

File: Dockerfile (hunk 1 above)
Comment (bradlarsen) [reply Dockerfile]: Thanks!

**Review by bradlarsen — APPROVED** [review]

**Pull request merged**